### PR TITLE
Calculate Vec capacities in librustc

### PR DESCRIPTION
**Pull request by ljedrz**
Calculate the required capacity of a few vectors in rustc based on the number of elements they are populated with.

### CODE REVIEW

**Comment by ljedrz** [comment]
Spurious error; please retry when GitHub is back on its feet.

**Comment by ljedrz** [comment]
Spurious error against.

**Comment by Mark-Simulacrum** [comment]
Retried. cc @nnethercote -- some of these are iterating over the data that'll be inserted into the vector to calculate the length; do you know off-hand if that's a bad idea or not?

**Comment by nnethercote** [comment]
> some of these are iterating over the data that'll be inserted into the vector to calculate the length; do you know off-hand if that's a bad idea or not?

Not sure. If the data will be iterated over subsequently and it's not too large, it probably doesn't matter. Profiling is the ultimate test.

(BTW, the PR lacks a description, as does the patch. That makes any kind of commentary more difficult. Even a single sentence of explanation would be helpful.)

**Comment by ljedrz** [comment]
@nnethercote I added a short description (initially I just didn't think it was too informative); please advise if I'm missing anything.

**Comment by nnethercote** [comment]
@ljedrz: Often the "why" is more interesting than the "what", especially for PRs and commit messages. Do you have measurements showing that this speeds things up?

**Comment by ljedrz** [comment]
I'd be happy to measure it; could you tell me which benchmarks I should run and how to do it? I couldn't find that information online.

**Comment by oli-obk** [comment]
@bors try

Let's do a perf run I guess?

**Comment by ljedrz** [comment]
I'll fix the error and do a perf run locally.

**Comment by ljedrz** [comment]
Hmm, I can't build rust-perf due to
```
error[E0425]: cannot find function `exec` in this scope
   --> collector\src/bin/rustc-fake.rs:139:9
    |
139 |         exec(&mut cmd);
    |         ^^^^ not found in this scope
```
I fixed the PR; @oli-obk can you try please?

**Comment by ljedrz** [comment]
`[00:18:42] error: could not parse/generate dep info at: /checkout/obj/build/x86_64-unknown-linux-gnu/stage1-std/x86_64-unknown-linux-gnu/release/deps/core-1cbcabaa1ea822b5.d` 😕?

**Comment by oli-obk** [comment]
Does `./x.py test` pass for you locally?

**Comment by ljedrz** [comment]
@oli-obk it does now; the culprit was `driver.rs` where for some reason the `keys` could not be accessed before the loop.

**Comment by ljedrz** [comment]
Is the latest proposal ok? I can always limit it to the simplest cases (like `astconv.rs`) if need be.

File: src/librustc_driver/driver.rs [hunk 1]
```diff
@@ -1353,7 +1353,8 @@ fn generated_output_paths(
     exact_name: bool,
     crate_name: &str,
 ) -> Vec<PathBuf> {
-    let mut out_filenames = Vec::new();
+    let mut out_filenames = Vec::with_capacity(sess.opts.output_types.len());
```
Comment (eddyb) [inline src/librustc_driver/driver.rs]: This doesn't appear to be correct.

File: src/librustc_typeck/astconv.rs [hunk 2]
```diff
@@ -1464,7 +1464,12 @@ impl<'a, 'gcx, 'tcx> Bounds<'tcx> {
     pub fn predicates(&self, tcx: TyCtxt<'a, 'gcx, 'tcx>, param_ty: Ty<'tcx>)
                       -> Vec<ty::Predicate<'tcx>>
     {
-        let mut vec = Vec::new();
+        let mut vec = Vec::with_capacity(
+            1 +
+            self.region_bounds.len() +
+            self.trait_bounds.len() +
+            self.projection_bounds.len()
+        );
```
Comment (eddyb) [inline src/librustc_typeck/astconv.rs]: This can be created from an iterator and `.collect()`.

**Comment by eddyb** [comment]
In general we should *really avoid* `Vec::with_capacity`, and `push` calls, if we can use iterators.

File: src/librustc_driver/driver.rs (hunk 1 above)
Comment (ljedrz) [reply src/librustc_driver/driver.rs]: It's later shrunk at 1380; processing `keys()` before the loop (which was in my initial commit) caused a really uninformative error to occur during stage 1 build.

**Comment by oli-obk** [comment]
@eddyb do you want these to be refactored to iterators + collect even if perf shows no benefits?

Otherwise I say we do a perf run and see if there's any point to doing this at all.

File: src/librustc_driver/driver.rs (hunk 1 above)
Comment (eddyb) [reply src/librustc_driver/driver.rs]: Then it's a better idea not to touch it at all. This is in the driver, and runs once per compilation, AFAICT, so it's not really worth optimizing (unless you can rewrite it to an iterator, that is).

**Comment by ljedrz** [comment]
@eddyb I addressed your comment, does it look good now?

File: src/librustc/middle/dead.rs [hunk 3]
```diff
@@ -396,16 +396,12 @@ impl<'v, 'k, 'tcx> ItemLikeVisitor<'v> for LifeSeeder<'k, 'tcx> {
...
-
-    // Seed entry point
-    if let Some((id, _, _)) = *tcx.sess.entry_fn.borrow() {
-        worklist.push(id);
-    }
+                                      -> Vec<ast::NodeId>
+{
+    let worklist = access_levels.map.iter().map(|(&id, _)| id).chain(
+        // Seed entry point
+        (*tcx.sess.entry_fn.borrow()).and_then(|(id, _, _)| Some(id))
```
Comment (eddyb) [inline src/librustc/middle/dead.rs]: `.and_then(|x| Some(y))` is always `.map(|x| y)` AFAIK

Comment (eddyb) [reply src/librustc/middle/dead.rs]: Oh and you shouldn't need the `*`, using by-value `Option` methods will do it for you.

File: src/librustc_codegen_llvm/abi.rs [hunk 4]
```diff
@@ -565,7 +565,13 @@ impl<'a, 'tcx> FnTypeExt<'a, 'tcx> for FnType<'tcx, Ty<'tcx>> {
     }
 
     fn llvm_type(&self, cx: &CodegenCx<'a, 'tcx>) -> Type {
-        let mut llargument_tys = Vec::new();
+        let args_capacity = self.args.iter().fold(0, |acc, arg| acc +
+            if arg.pad.is_some() { 1 } else { 0 } +
+            if let PassMode::Pair(_, _) = arg.mode { 2 } else { 1 }
+        );
```
Comment (eddyb) [inline src/librustc_codegen_llvm/abi.rs]: You can use `map` and `sum`.

File: src/librustc_typeck/astconv.rs [hunk 5]
```diff
@@ -1464,35 +1464,34 @@ impl<'a, 'gcx, 'tcx> Bounds<'tcx> {
...
-        }
-
-        for bound_trait_ref in &self.trait_bounds {
-            vec.push(bound_trait_ref.to_predicate());
-        }
-
-        for projection in &self.projection_bounds {
-            vec.push(projection.to_predicate());
-        }
+                Some(trait_ref.to_predicate())
```
Comment (eddyb) [inline src/librustc_typeck/astconv.rs]: Same thing here, `and_then` can be just `map`.

**Comment by kennytm** [comment]
@rust-timer build 435ce0bad1c0bc79ec8f0591eba110a92492e6ff

**Comment by ljedrz** [comment]
@eddyb done.

File: src/librustc/middle/dead.rs [hunk 6]
```diff
@@ -396,16 +396,12 @@ impl<'v, 'k, 'tcx> ItemLikeVisitor<'v> for LifeSeeder<'k, 'tcx> {
...
-
-    // Seed entry point
-    if let Some((id, _, _)) = *tcx.sess.entry_fn.borrow() {
-        worklist.push(id);
-    }
+                                      -> Vec<ast::NodeId>
+{
+    let worklist = access_levels.map.iter().map(|(&id, _)| id).chain(
+        // Seed entry point
+        (tcx.sess.entry_fn.borrow()).map(|(id, _, _)| id)
```
[the remark this answers is not in the document]
Comment (ljedrz) [reply src/librustc/middle/dead.rs]: You're right, I should've noticed that; done.

**Pull request merged**